adapters/repos/db/vector/composer/index.go

```diff
@@ -0,0 +1,453 @@
+	DistanceBetweenVectors(x, y []float32) (float32, bool, error)
+	ContainsNode(id uint64) bool
+	DistancerProvider() distancer.Provider
+	AlreadyIndexed() uint64
+}
+
+type upgradableIndexer interface {
+	Upgraded() bool
+	Upgrade(callback func()) error
+	ShouldUpgrade() (bool, int)
+}
+
+type composer struct {
+	sync.RWMutex
+	id                       string
+	targetVector             string
+	store                    *lsmkv.Store
+	logger                   logrus.FieldLogger
+	rootPath                 string
+	shardName                string
+	className                string
+	prometheusMetrics        *monitoring.PrometheusMetrics
+	vectorForIDThunk         common.VectorForID[float32]
+	tempVectorForIDThunk     common.TempVectorForID
+	distanceProvider         distancer.Provider
+	makeCommitLoggerThunk    hnsw.MakeCommitLogger
+	threshold                uint64
+	index                    VectorIndex
+	upgraded                 atomic.Bool
+	tombstoneCallbacks       cyclemanager.CycleCallbackGroup
+	shardCompactionCallbacks cyclemanager.CycleCallbackGroup
+	shardFlushCallbacks      cyclemanager.CycleCallbackGroup
+	hnswUC                   hnswent.UserConfig
+	db                       *bolt.DB
+}
+
+func New(cfg Config, uc ent.UserConfig, store *lsmkv.Store) (*composer, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid config")
+	}
+
+	logger := cfg.Logger
+	if logger == nil {
+		l := logrus.New()
+		l.Out = io.Discard
+		logger = l
+	}
+
+	flatConfig := flat.Config{
+		ID:               cfg.ID,
+		TargetVector:     cfg.TargetVector,
+		Logger:           cfg.Logger,
+		DistanceProvider: cfg.DistanceProvider,
+	}
+
+	index := &composer{
+		id:                       cfg.ID,
+		targetVector:             cfg.TargetVector,
+		logger:                   logger,
+		rootPath:                 cfg.RootPath,
+		shardName:                cfg.ShardName,
+		className:                cfg.ClassName,
+		prometheusMetrics:        cfg.PrometheusMetrics,
+		vectorForIDThunk:         cfg.VectorForIDThunk,
+		tempVectorForIDThunk:     cfg.TempVectorForIDThunk,
+		distanceProvider:         cfg.DistanceProvider,
+		makeCommitLoggerThunk:    cfg.MakeCommitLoggerThunk,
+		store:                    store,
+		threshold:                uc.Threeshold,
+		tombstoneCallbacks:       cfg.TombstoneCallbacks,
+		shardCompactionCallbacks: cfg.ShardCompactionCallbacks,
+		shardFlushCallbacks:      cfg.ShardFlushCallbacks,
+		hnswUC:                   uc.HnswUC,
+	}
+
+	path := filepath.Join(cfg.RootPath, "index.db")
+
+	db, err := bolt.Open(path, 0o600, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "open %q", path)
+	}
+	err = db.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(composerBucket)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	upgraded := false
+	err = db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket(composerBucket)
+		v := b.Get([]byte{0})
```

Comment:
We should probably choose a more explicit key. Example:

```go

const composerUpgradedKey = "upgraded"
...
b.Put([]byte(composerUpgradedKey), []byte{1})
```

Reply:
I was not able to merge anymore since there were more changes. I did commit the proposed changes to the final PR though. Thanks.